irremote/irremote.go

```diff
@@ -0,0 +1,173 @@
+package irremote // import "tinygo.org/x/drivers/irremote"
+
+import (
+	"machine"
+	"time"
+)
+
+// NEC protocol references
+// https://www.sbprojects.net/knowledge/ir/nec.php
+// https://techdocs.altium.com/display/FPGA/NEC+Infrared+Transmission+Protocol
+// https://simple-circuit.com/arduino-nec-remote-control-decoder/
+
+// Client callback function for IR commands
+// 'code' is the raw NEC format code. 'addr' & 'cmd' are the NEC decoded address and command codes respectively
+// See the NEC protocol reference links above for more information
+type IRCallback func(code uint32, addr uint16, cmd uint8, repeat bool)
+
+// Public API for an IR receiver device
+type IRReceiverDevice interface {
+	// Configure pins for the IR receiver module
+	Configure()
+	// Start/Stop receiving IR callbacks (pass nil to stop)
+	Callback(cb IRCallback)
```

Comment:
Perhaps better to name this function something like `HandleDataComplete()`?

Reply:
I agree that this is badly named and needs changing, although I'm not too keen on `HandleDataComplete()` (sorry!)  Personally I'd prefer to avoid the terminology of 'complete' in this case. My justification for this is that multiple IR commands/codes occur in discrete packets over time with no real concept of 'completing' or 'terminating' until you set the callback to `nil` to tell the driver you're not interested in them anymore. i.e. it's not a continous stream that has a natural completion. Neither is it a 'one-shot' async call that 'completes' Not sure I'm describring that too well but hopefully you can understand my thoughts?

It's a 'setter' function - you are setting a callback function property, passing `nil` to clear it - so IMHO it should be named something of the form `SetXxxYyy()` How about  `SetCommandHandler()`, `SetCommandCallback()`,  or just plain `SetCallback()`??

On a related note I was questioning the use of using callback functions from an interrupt handler. I'm not sure if this is a good idea? I've noticed on Pico that any call to `time.Sleep()` in an interrupt handler usually kills the program (although I'm not sure if this is a bug?) However, the only alternative I can think of is to use a channel to 'transpose' the callback back to the client thread/goroutine. But use of channels would enforce the use of a scheduler which isn't great for the more constrained devices like AVR. Any insight/advice here would be gratefully received! Thanks!